Support `"file://…"` for paths passed to `dlopen(…)` (from `bun:ffi`)

This is a new issue based on https://github.com/oven-sh/bun/issues/9878#issuecomment-2039068136

### What version of Bun is running?

1.1.4

### What platform is your computer?

Darwin 23.4.0 arm64 arm

### What steps can reproduce the bug?

```shell
git clone https://github.com/cubing/twsearch && cd twsearch
git checkout 6279f4abbee4ac8941d7cd5648c6811eb283cb0a
cargo install cargo-run-bin # Requires `cargo` to be installed already.
make build-rust-ffi
bun run "src/rs-ffi/test/js_test.ts"
```

### What is the expected behavior?

`dlopen` opens the `file://` path from `import.meta.resolve(…)`

### What do you see instead?

```
 8 |   `../../../target/release/libtwsearch_ffi.${suffix}`,
 9 | );
10 | 
11 | const {
12 |   symbols: { ffi_random_scramble_for_event },
13 | } = dlopen(path, {
         ^
ERR_DLOPEN_FAILED: Failed to open library. This is usually caused by a missing library or an invalid library path.
 syscall: "dlopen"

      at bun:ffi:61:41
      at /Users/lgarron/Code/git/github.com/cubing/twsearch/src/rs-ffi/test/js_test.ts:13:5
make: *** [test-rust-ffi-js] Error 1
```

### Additional information

I was glad to see https://github.com/oven-sh/bun/pull/9883 land in `bun` 1.1.1, but it doesn't seem to work on my computer at all. 🥺
I would love to be able to use it!

--- a/packages/bun-types/ffi.d.ts
+++ b/packages/bun-types/ffi.d.ts
@@ -592,7 +592,10 @@ declare module "bun:ffi" {
    * bun uses [tinycc](https://github.com/TinyCC/tinycc), so a big thanks
    * goes to Fabrice Bellard and TinyCC maintainers for making this possible.
    */
-  function dlopen<Fns extends Record<string, FFIFunction>>(name: string, symbols: Fns): Library<Fns>;
+  function dlopen<Fns extends Record<string, FFIFunction>>(
+    name: string | import("bun").BunFile | URL,
+    symbols: Fns,
+  ): Library<Fns>;
 
   /**
    * Turn a native library's function pointer into a JavaScript function

--- a/src/bun.js/api/ffi.zig
+++ b/src/bun.js/api/ffi.zig
@@ -273,19 +273,39 @@ pub const FFI = struct {
 
     pub fn open(global: *JSGlobalObject, name_str: ZigString, object: JSC.JSValue) JSC.JSValue {
         JSC.markBinding(@src());
-        const allocator = VirtualMachine.get().allocator;
+        const vm = VirtualMachine.get();
+        const allocator = bun.default_allocator;
         var name_slice = name_str.toSlice(allocator);
         defer name_slice.deinit();
-
-        if (name_slice.len == 0) {
-            return JSC.toInvalidArguments("Invalid library name", .{}, global);
-        }
 
         if (object.isEmptyOrUndefinedOrNull() or !object.isObject()) {
             return JSC.toInvalidArguments("Expected an options object with symbol names", .{}, global);
         }
 
-        const name = name_slice.slice();
+        var filepath_buf: [bun.MAX_PATH_BYTES]u8 = undefined;
+        const name = brk: {
+            if (JSC.ModuleLoader.resolveEmbeddedFile(
+                vm,
+                name_slice.slice(),
+                switch (Environment.os) {
+                    .linux => "so",
+                    .mac => "dylib",
+                    .windows => "dll",
+                    else => @compileError("TODO"),
+                },
+            )) |resolved| {
+                @memcpy(filepath_buf[0..resolved.len], resolved);
+                filepath_buf[resolved.len] = 0;
+                break :brk filepath_buf[0..resolved.len];
+            }
+
+            break :brk name_slice.slice();
+        };
+
+        if (name.len == 0) {
+            return JSC.toInvalidArguments("Invalid library name", .{}, global);
+        }
+
         var symbols = bun.StringArrayHashMapUnmanaged(Function){};
         if (generateSymbols(global, &symbols, object) catch JSC.JSValue.zero) |val| {
             // an error while validating symbols
@@ -329,7 +349,7 @@ pub const FFI = struct {
             // optional if the user passed "ptr"
             if (function.symbol_from_dynamic_library == null) {
                 const resolved_symbol = dylib.lookup(*anyopaque, function_name) orelse {
-                    const ret = JSC.toInvalidArguments("Symbol \"{s}\" not found in \"{s}\"", .{ bun.asByteSlice(function_name), name_slice.slice() }, global);
+                    const ret = JSC.toInvalidArguments("Symbol \"{s}\" not found in \"{s}\"", .{ bun.asByteSlice(function_name), name }, global);
                     for (symbols.values()) |*value| {
                         allocator.free(@constCast(bun.asByteSlice(value.base_name.?)));
                         value.arg_types.clearAndFree(allocator);
@@ -346,7 +366,7 @@ pub const FFI = struct {
                 const ret = JSC.toInvalidArguments("{s} when compiling symbol \"{s}\" in \"{s}\"", .{
                     bun.asByteSlice(@errorName(err)),
                     bun.asByteSlice(function_name),
-                    name_slice.slice(),
+                    name,
                 }, global);
                 for (symbols.values()) |*value| {
                     allocator.free(@constCast(bun.asByteSlice(value.base_name.?)));

--- a/src/bun.js/module_loader.zig
+++ b/src/bun.js/module_loader.zig
@@ -721,6 +721,49 @@ pub const ModuleLoader = struct {
         }
     }
 
+    pub fn resolveEmbeddedFile(vm: *JSC.VirtualMachine, input_path: []const u8, extname: []const u8) ?[]const u8 {
+        if (input_path.len == 0) return null;
+        var graph = vm.standalone_module_graph orelse return null;
+        const file = graph.find(input_path) orelse return null;
+
+        if (comptime Environment.isLinux) {
+            // TODO: use /proc/fd/12346 instead! Avoid the copy!
+        }
+
+        // atomically write to a tmpfile and then move it to the final destination
+        var tmpname_buf: [bun.MAX_PATH_BYTES]u8 = undefined;
+        const tmpfilename = bun.sliceTo(bun.fs.FileSystem.instance.tmpname(extname, &tmpname_buf, bun.hash(file.name)) catch return null, 0);
+
+        const tmpdir = bun.fs.FileSystem.instance.tmpdir() catch return null;
+
+        // First we open the tmpfile, to avoid any other work in the event of failure.
+        const tmpfile = bun.Tmpfile.create(bun.toFD(tmpdir.fd), tmpfilename).unwrap() catch return null;
+        defer {
+            _ = bun.sys.close(tmpfile.fd);
+        }
+
+        switch (JSC.Node.NodeFS.writeFileWithPathBuffer(
+            &tmpname_buf, // not used
+
+            .{
+                .data = .{
+                    .encoded_slice = JSC.ZigString.Slice.fromUTF8NeverFree(file.contents),
+                },
+                .dirfd = bun.toFD(tmpdir.fd),
+                .file = .{
+                    .fd = tmpfile.fd,
+                },
+                .encoding = .buffer,
+            },
+        )) {
+            .err => {
+                return null;
+            },
+            else => {},
+        }
+        return bun.path.joinAbs(bun.fs.FileSystem.instance.fs.tmpdirPath(), .auto, tmpfilename);
+    }
+
     pub const AsyncModule = struct {
 
         // This is all the state used by the printer to print the module
@@ -2911,47 +2954,11 @@ pub const HardcodedModule = enum {
 
 /// Support embedded .node files
 export fn Bun__resolveEmbeddedNodeFile(vm: *JSC.VirtualMachine, in_out_str: *bun.String) bool {
-    var graph = vm.standalone_module_graph orelse return false;
-    const utf8 = in_out_str.toUTF8(bun.default_allocator);
-    defer utf8.deinit();
-    const file = graph.find(utf8.slice()) orelse return false;
+    if (vm.standalone_module_graph == null) return false;
 
-    if (comptime Environment.isLinux) {
-        // TODO: use /proc/fd/12346 instead! Avoid the copy!
-    }
-
-    // atomically write to a tmpfile and then move it to the final destination
-    var tmpname_buf: [bun.MAX_PATH_BYTES]u8 = undefined;
-    const tmpfilename = bun.sliceTo(bun.fs.FileSystem.instance.tmpname("node", &tmpname_buf, bun.hash(file.name)) catch return false, 0);
-
-    const tmpdir = bun.fs.FileSystem.instance.tmpdir() catch return false;
-
-    // First we open the tmpfile, to avoid any other work in the event of failure.
-    const tmpfile = bun.Tmpfile.create(bun.toFD(tmpdir.fd), tmpfilename).unwrap() catch return false;
-    defer {
-        _ = bun.sys.close(tmpfile.fd);
-    }
-
-    switch (JSC.Node.NodeFS.writeFileWithPathBuffer(
-        &tmpname_buf, // not used
-
-        .{
-            .data = .{
-                .encoded_slice = JSC.ZigString.Slice.fromUTF8NeverFree(file.contents),
-            },
-            .dirfd = bun.toFD(tmpdir.fd),
-            .file = .{
-                .fd = tmpfile.fd,
-            },
-            .encoding = .buffer,
-        },
-    )) {
-        .err => {
-            return false;
-        },
-        else => {},
-    }
-
-    in_out_str.* = bun.String.createUTF8(bun.path.joinAbs(bun.fs.FileSystem.instance.fs.tmpdirPath(), .auto, tmpfilename));
+    const input_path = in_out_str.toUTF8(bun.default_allocator);
+    defer input_path.deinit();
+    const result = ModuleLoader.resolveEmbeddedFile(vm, input_path.slice(), "node") orelse return false;
+    in_out_str.* = bun.String.createUTF8(result);
     return true;
 }

--- a/src/js/bun/ffi.ts
+++ b/src/js/bun/ffi.ts
@@ -396,6 +396,22 @@ const native = {
 };
 
 function dlopen(path, options) {
+  if (typeof path === "string" && path?.startsWith?.("file:")) {
+    // import.meta.url returns a file: URL
+    // https://github.com/oven-sh/bun/issues/10304
+    path = Bun.fileURLToPath(path);
+  } else if (typeof path === "object" && path) {
+    if (path instanceof URL) {
+      // This is mostly for import.meta.resolve()
+      // https://github.com/oven-sh/bun/issues/10304
+      path = Bun.fileURLToPath(path as URL);
+    } else if (path instanceof Blob) {
+      // must be a Bun.file() blob
+      // https://discord.com/channels/876711213126520882/1230114905898614794/1230114905898614794
+      path = path.name;
+    }
+  }
+
   const result = nativeDLOpen(path, options);
   if (result instanceof Error) throw result;
 

--- a/test/js/bun/ffi/ffi.test.js
+++ b/test/js/bun/ffi/ffi.test.js
@@ -680,280 +680,298 @@ const lib_path =
     : existsSync("/lib/x86_64-linux-gnu/libc.so.6")
       ? "/lib/x86_64-linux-gnu/libc.so.6"
       : null;
-const test = lib_path ? it : it.skip;
-test("can open more than 63 symbols", () => {
-  const lib = dlopen(lib_path, {
-    memchr: {
-      returns: "ptr",
-      args: ["ptr", "int", "usize"],
-    },
-    strcpy: {
-      returns: "ptr",
-      args: ["ptr", "ptr"],
-    },
-    strcat: {
-      returns: "ptr",
-      args: ["ptr", "ptr"],
-    },
-    strncat: {
-      returns: "ptr",
-      args: ["ptr", "ptr", "usize"],
-    },
-    strcmp: {
-      returns: "int",
-      args: ["ptr", "ptr"],
-    },
-    strncmp: {
-      returns: "int",
-      args: ["ptr", "ptr", "usize"],
-    },
-    strcoll: {
-      returns: "int",
-      args: ["ptr", "ptr"],
-    },
-    strxfrm: {
-      returns: "int",
-      args: ["ptr", "ptr", "usize"],
-    },
-    strchr: {
-      returns: "ptr",
-      args: ["ptr", "int"],
-    },
-    strrchr: {
-      returns: "ptr",
-      args: ["ptr", "int"],
-    },
-    strcspn: {
-      returns: "usize",
-      args: ["ptr", "ptr"],
-    },
-    strspn: {
-      returns: "usize",
-      args: ["ptr", "ptr"],
-    },
-    strpbrk: {
-      returns: "ptr",
-      args: ["ptr", "ptr"],
-    },
-    strstr: {
-      returns: "ptr",
-      args: ["ptr", "ptr"],
-    },
-    strtok: {
-      returns: "ptr",
-      args: ["ptr", "ptr"],
-    },
-    strerror: {
-      returns: "ptr",
-      args: ["int"],
-    },
-    strerror_r: {
-      returns: "ptr",
-      args: ["int", "ptr", "usize"],
-    },
-    strsep: {
-      returns: "ptr",
-      args: ["ptr", "ptr"],
-    },
-    strsignal: {
-      returns: "ptr",
-      args: ["int"],
-    },
-    stpcpy: {
-      returns: "ptr",
-      args: ["ptr", "ptr"],
-    },
-    stpncpy: {
-      returns: "ptr",
-      args: ["ptr", "ptr", "usize"],
-    },
-    basename: {
-      returns: "ptr",
-      args: ["ptr"],
-    },
-    bcmp: {
-      returns: "int",
-      args: ["ptr", "ptr", "usize"],
-    },
-    getdate: {
-      returns: "ptr",
-      args: ["ptr"],
-    },
-    gmtime: {
-      returns: "ptr",
-      args: ["ptr"],
-    },
-    localtime: {
-      returns: "ptr",
-      args: ["ptr"],
-    },
-    ctime: {
-      returns: "ptr",
-      args: ["ptr"],
-    },
-    asctime: {
-      returns: "ptr",
-      args: ["ptr"],
-    },
-    strftime: {
-      returns: "usize",
-      args: ["ptr", "usize", "ptr", "ptr"],
-    },
-    strptime: {
-      returns: "ptr",
-      args: ["ptr", "ptr", "ptr"],
-    },
-    asctime_r: {
-      returns: "ptr",
-      args: ["ptr", "ptr"],
-    },
-    ctime_r: {
-      returns: "ptr",
-      args: ["ptr", "ptr"],
-    },
-    gmtime_r: {
-      returns: "ptr",
-      args: ["ptr", "ptr"],
-    },
-    localtime_r: {
-      returns: "ptr",
-      args: ["ptr", "ptr"],
-    },
-    bcopy: {
-      returns: "int",
-      args: ["ptr", "ptr", "usize"],
-    },
-    bzero: {
-      returns: "void",
-      args: ["ptr", "usize"],
-    },
-    index: {
-      returns: "ptr",
-      args: ["ptr", "int"],
-    },
-    rindex: {
-      returns: "ptr",
-      args: ["ptr", "int"],
-    },
-    ffs: {
-      returns: "int",
-      args: ["int"],
-    },
-    strcasecmp: {
-      returns: "int",
-      args: ["ptr", "ptr"],
-    },
-    strncasecmp: {
-      returns: "int",
-      args: ["ptr", "ptr", "usize"],
-    },
-    pthread_attr_init: {
-      returns: "int",
-      args: ["ptr"],
-    },
-    pthread_attr_destroy: {
-      returns: "int",
-      args: ["ptr"],
-    },
-    pthread_attr_getdetachstate: {
-      returns: "int",
-      args: ["ptr", "ptr"],
-    },
-    pthread_attr_setdetachstate: {
-      returns: "int",
-      args: ["ptr", "int"],
-    },
-    pthread_attr_getguardsize: {
-      returns: "int",
-      args: ["ptr", "ptr"],
-    },
-    pthread_attr_setguardsize: {
-      returns: "int",
-      args: ["ptr", "usize"],
-    },
-    pthread_attr_getschedparam: {
-      returns: "int",
-      args: ["ptr", "ptr"],
-    },
-    pthread_attr_setschedparam: {
-      returns: "int",
-      args: ["ptr", "ptr"],
-    },
-    pthread_attr_getschedpolicy: {
-      returns: "int",
-      args: ["ptr", "ptr"],
-    },
-    pthread_attr_setschedpolicy: {
-      returns: "int",
-      args: ["ptr", "int"],
-    },
-    pthread_attr_getinheritsched: {
-      returns: "int",
-      args: ["ptr", "ptr"],
-    },
-    pthread_attr_setinheritsched: {
-      returns: "int",
-      args: ["ptr", "int"],
-    },
-    pthread_attr_getscope: {
-      returns: "int",
-      args: ["ptr", "ptr"],
-    },
-    pthread_attr_setscope: {
-      returns: "int",
-      args: ["ptr", "int"],
-    },
-    pthread_attr_getstackaddr: {
-      returns: "int",
-      args: ["ptr", "ptr"],
-    },
-    pthread_attr_setstackaddr: {
-      returns: "int",
-      args: ["ptr", "ptr"],
-    },
-    pthread_attr_getstacksize: {
-      returns: "int",
-      args: ["ptr", "ptr"],
-    },
-    pthread_attr_setstacksize: {
-      returns: "int",
-      args: ["ptr", "usize"],
-    },
-    pthread_attr_getstack: {
-      returns: "int",
-      args: ["ptr", "ptr", "ptr"],
-    },
-    pthread_attr_setstack: {
-      returns: "int",
-      args: ["ptr", "ptr", "usize"],
-    },
-    pthread_attr_getguardsize: {
-      returns: "int",
-      args: ["ptr", "ptr"],
-    },
-    pthread_attr_setguardsize: {
-      returns: "int",
-      args: ["ptr", "usize"],
-    },
-    login_tty: {
-      returns: "int",
-      args: ["int"],
-    },
-    login: {
-      returns: "int",
-      args: ["ptr"],
-    },
-    logout: {
-      returns: "int",
-      args: ["ptr"],
-    },
-    strlen: {
-      returns: "usize",
-      args: ["ptr"],
-    },
-  });
+const test = lib_path ? describe : describe.skip;
+test("can open more than 63 symbols via", () => {
+  if (lib_path) {
+    for (const [libPath, description] of [
+      // For file: URLs since one might do import.meta.resolve()
+      [Bun.pathToFileURL(lib_path), "URL"],
 
-  expect(Object.keys(lib.symbols).length).toBe(65);
-  expect(lib.symbols.strcasecmp(Buffer.from("ciro\0"), Buffer.from("CIRO\0"))).toBe(0);
-  expect(lib.symbols.strlen(Buffer.from("bunbun\0", "ascii"))).toBe(6n);
+      // file: URLs as a string
+      [Bun.pathToFileURL(lib_path).href, "file: URL"],
+
+      // For embedding files since one might do Bun.file(embeddedFile)
+      [Bun.file(lib_path), "Bun.file"],
+
+      // For file path strings
+      [lib_path, "string"],
+    ]) {
+      it(description, () => {
+        // test file: URL works
+        const lib = dlopen(libPath, {
+          memchr: {
+            returns: "ptr",
+            args: ["ptr", "int", "usize"],
+          },
+          strcpy: {
+            returns: "ptr",
+            args: ["ptr", "ptr"],
+          },
+          strcat: {
+            returns: "ptr",
+            args: ["ptr", "ptr"],
+          },
+          strncat: {
+            returns: "ptr",
+            args: ["ptr", "ptr", "usize"],
+          },
+          strcmp: {
+            returns: "int",
+            args: ["ptr", "ptr"],
+          },
+          strncmp: {
+            returns: "int",
+            args: ["ptr", "ptr", "usize"],
+          },
+          strcoll: {
+            returns: "int",
+            args: ["ptr", "ptr"],
+          },
+          strxfrm: {
+            returns: "int",
+            args: ["ptr", "ptr", "usize"],
+          },
+          strchr: {
+            returns: "ptr",
+            args: ["ptr", "int"],
+          },
+          strrchr: {
+            returns: "ptr",
+            args: ["ptr", "int"],
+          },
+          strcspn: {
+            returns: "usize",
+            args: ["ptr", "ptr"],
+          },
+          strspn: {
+            returns: "usize",
+            args: ["ptr", "ptr"],
+          },
+          strpbrk: {
+            returns: "ptr",
+            args: ["ptr", "ptr"],
+          },
+          strstr: {
+            returns: "ptr",
+            args: ["ptr", "ptr"],
+          },
+          strtok: {
+            returns: "ptr",
+            args: ["ptr", "ptr"],
+          },
+          strerror: {
+            returns: "ptr",
+            args: ["int"],
+          },
+          strerror_r: {
+            returns: "ptr",
+            args: ["int", "ptr", "usize"],
+          },
+          strsep: {
+            returns: "ptr",
+            args: ["ptr", "ptr"],
+          },
+          strsignal: {
+            returns: "ptr",
+            args: ["int"],
+          },
+          stpcpy: {
+            returns: "ptr",
+            args: ["ptr", "ptr"],
+          },
+          stpncpy: {
+            returns: "ptr",
+            args: ["ptr", "ptr", "usize"],
+          },
+          basename: {
+            returns: "ptr",
+            args: ["ptr"],
+          },
+          bcmp: {
+            returns: "int",
+            args: ["ptr", "ptr", "usize"],
+          },
+          getdate: {
+            returns: "ptr",
+            args: ["ptr"],
+          },
+          gmtime: {
+            returns: "ptr",
+            args: ["ptr"],
+          },
+          localtime: {
+            returns: "ptr",
+            args: ["ptr"],
+          },
+          ctime: {
+            returns: "ptr",
+            args: ["ptr"],
+          },
+          asctime: {
+            returns: "ptr",
+            args: ["ptr"],
+          },
+          strftime: {
+            returns: "usize",
+            args: ["ptr", "usize", "ptr", "ptr"],
+          },
+          strptime: {
+            returns: "ptr",
+            args: ["ptr", "ptr", "ptr"],
+          },
+          asctime_r: {
+            returns: "ptr",
+            args: ["ptr", "ptr"],
+          },
+          ctime_r: {
+            returns: "ptr",
+            args: ["ptr", "ptr"],
+          },
+          gmtime_r: {
+            returns: "ptr",
+            args: ["ptr", "ptr"],
+          },
+          localtime_r: {
+            returns: "ptr",
+            args: ["ptr", "ptr"],
+          },
+          bcopy: {
+            returns: "int",
+            args: ["ptr", "ptr", "usize"],
+          },
+          bzero: {
+            returns: "void",
+            args: ["ptr", "usize"],
+          },
+          index: {
+            returns: "ptr",
+            args: ["ptr", "int"],
+          },
+          rindex: {
+            returns: "ptr",
+            args: ["ptr", "int"],
+          },
+          ffs: {
+            returns: "int",
+            args: ["int"],
+          },
+          strcasecmp: {
+            returns: "int",
+            args: ["ptr", "ptr"],
+          },
+          strncasecmp: {
+            returns: "int",
+            args: ["ptr", "ptr", "usize"],
+          },
+          pthread_attr_init: {
+            returns: "int",
+            args: ["ptr"],
+          },
+          pthread_attr_destroy: {
+            returns: "int",
+            args: ["ptr"],
+          },
+          pthread_attr_getdetachstate: {
+            returns: "int",
+            args: ["ptr", "ptr"],
+          },
+          pthread_attr_setdetachstate: {
+            returns: "int",
+            args: ["ptr", "int"],
+          },
+          pthread_attr_getguardsize: {
+            returns: "int",
+            args: ["ptr", "ptr"],
+          },
+          pthread_attr_setguardsize: {
+            returns: "int",
+            args: ["ptr", "usize"],
+          },
+          pthread_attr_getschedparam: {
+            returns: "int",
+            args: ["ptr", "ptr"],
+          },
+          pthread_attr_setschedparam: {
+            returns: "int",
+            args: ["ptr", "ptr"],
+          },
+          pthread_attr_getschedpolicy: {
+            returns: "int",
+            args: ["ptr", "ptr"],
+          },
+          pthread_attr_setschedpolicy: {
+            returns: "int",
+            args: ["ptr", "int"],
+          },
+          pthread_attr_getinheritsched: {
+            returns: "int",
+            args: ["ptr", "ptr"],
+          },
+          pthread_attr_setinheritsched: {
+            returns: "int",
+            args: ["ptr", "int"],
+          },
+          pthread_attr_getscope: {
+            returns: "int",
+            args: ["ptr", "ptr"],
+          },
+          pthread_attr_setscope: {
+            returns: "int",
+            args: ["ptr", "int"],
+          },
+          pthread_attr_getstackaddr: {
+            returns: "int",
+            args: ["ptr", "ptr"],
+          },
+          pthread_attr_setstackaddr: {
+            returns: "int",
+            args: ["ptr", "ptr"],
+          },
+          pthread_attr_getstacksize: {
+            returns: "int",
+            args: ["ptr", "ptr"],
+          },
+          pthread_attr_setstacksize: {
+            returns: "int",
+            args: ["ptr", "usize"],
+          },
+          pthread_attr_getstack: {
+            returns: "int",
+            args: ["ptr", "ptr", "ptr"],
+          },
+          pthread_attr_setstack: {
+            returns: "int",
+            args: ["ptr", "ptr", "usize"],
+          },
+          pthread_attr_getguardsize: {
+            returns: "int",
+            args: ["ptr", "ptr"],
+          },
+          pthread_attr_setguardsize: {
+            returns: "int",
+            args: ["ptr", "usize"],
+          },
+          login_tty: {
+            returns: "int",
+            args: ["int"],
+          },
+          login: {
+            returns: "int",
+            args: ["ptr"],
+          },
+          logout: {
+            returns: "int",
+            args: ["ptr"],
+          },
+          strlen: {
+            returns: "usize",
+            args: ["ptr"],
+          },
+        });
+        expect(Object.keys(lib.symbols).length).toBe(65);
+        expect(lib.symbols.strcasecmp(Buffer.from("ciro\0"), Buffer.from("CIRO\0"))).toBe(0);
+        expect(lib.symbols.strlen(Buffer.from("bunbun\0", "ascii"))).toBe(6n);
+      });
+    }
+  }
 });

